js/src/Map.js

```diff
@@ -6,6 +6,21 @@ var LeafletWidgetView = utils.LeafletWidgetView;
 var LeafletDOMWidgetView = utils.LeafletDOMWidgetView;
 var def_loc = [0.0, 0.0];
 
+var LeafletMapStyleModel = widgets.StyleModel.extend({
+    defaults: _.extend({}, widgets.StyleModel.prototype.defaults, {
+        _model_name : "LeafletMapStyleModel",
+        _model_module : "jupyter-leaflet",
+        }),
+
+    styleProperties: {
```

Comment:
I'm wondering if this is the right syntax

Reply:
This is how it looks in `ipywidgets`:

``` typescript
export
class ButtonStyleModel extends StyleModel {
    defaults() {
        return _.extend(super.defaults(), {
            _model_name: 'ButtonStyleModel',
            _model_module: '@jupyter-widgets/controls',
            _model_module_version: JUPYTER_CONTROLS_VERSION,

        });
    }

    public static styleProperties = {
        button_color: {
            selector: '',
            attribute: 'background-color',
            default: null
        },
        font_weight: {
            selector: '',
            attribute: 'font-weight',
            default: ''
        }
    };
}
```

Reply:
I'm out of the office now.. I'll get back in about 15 hours.. but, just before leaving the office I tried with the compiled version of `ButtonStyleModel` I found in `webpack:///packages/controls/lib/widget_button.js`. I replaced the `LeafletMapStyleModel` in the js side with this:

``` javascript
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var LeafletMapStyleModel = /** @class */ (function (_super) {
    __extends(LeafletMapStyleModel, _super);
    function LeafletMapStyleModel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    LeafletMapStyleModel.prototype.defaults = function () {
        return _.extend(_super.prototype.defaults.call(this), {
            _model_name: 'LeafletMapStyleModel',
            _model_module: 'jupyter-leaflet',
        });
    };
    LeafletMapStyleModel.styleProperties = {
        cursor: {
            selector: '',
            attribute: 'cursor',
            default: 'move'
        }
    };
    return LeafletMapStyleModel;
}(widgets.StyleModel));
```
and it worked! =)
But I cannot see what's the difference that makes this work and the other not.. :thinking:

Reply:
Yes.. I guess you can just do:
```javascript
    LeafletMapStyleModel.styleProperties = {
        cursor: {
            selector: '',
            attribute: 'cursor',
            default: 'move'
        }
    };
```
Just after the `LeafletMapStyleModel` definition. That should work. At some point I would like ipyleaflet to be written in TypeScript.

Reply:
I am testing now.. and have seen that it works when `selector: ''`, but if I try to change another selector style it doesn't work, for example `selector: '.leaflet-grab'`. The thing is that with the option that works, the cursor stays the same in the dragging process, and to return that behavior back I have to set cursor's style to `''`. Maybe `Style` widget cannot change css class properties. Do you think is still good?

Reply:
I don't know how it works honestly. I'm not sure I get why it works with `selector: ''` and not with `selector: '.leaflet-grab'`. I don't think the `Style` widget changes the css class properties, only the element style attributes. There is also the `.leaflet-dragging` class that you can use as a selector for when the user is dragging the `Map`, maybe that could be useful.

Reply:
Ok, I think I know why it does not work with the selector. That's because the selector is used here https://github.com/jupyter-widgets/ipywidgets/blob/1f93ad39802d3ee4a6b88bc13064f93f8f796374/packages/base/src/widget_style.ts#L72. The queryselector is used directly from the `el` attribute (the map itself). Just keep it without the selector we can fix it later.

Reply:
yes, but this approach cannot change any class selector, like `.leaflet-grab`, `.leaflet-container` or `.leaflet-dragging`... to me it'd be still very useful to change the cursor this way.. but I don't know if everyone agrees.
I leave some gifs that explain better:

## Changing cursor's type using a css class attribute
![drag](https://user-images.githubusercontent.com/10751505/56736057-8e534a80-673d-11e9-857e-ff5fd0d1ac76.gif)

## Changing cursor's type using the element css attribute (this approach)
![nodrag](https://user-images.githubusercontent.com/10751505/56736373-5bf61d00-673e-11e9-9eed-825ff5dd6170.gif)

Reply:
I think a way for fixing this later will be to put the `Map` inside of a subcontainer under the `el` element.

Reply:
Ok, do you mind me pushing to your branch?

Reply:
Not at all! I'm so excited I can contribute to this great project =)

Reply:
I opened a PR on your fork https://github.com/fitoprincipe/ipyleaflet/pull/1 :) It should fix your selector issue if you merge it on your branch :)

Reply:
Mmm.. works as before.. changes the element's style.. not `.leaflet-grab` style.. and when dragging, the cursor stays the same :thinking:

Reply:
It cannot change the `.leaflet-grab` class. AFAIK you cannot achieve that in JavaScript. The only thing you can do is select an element given a condition (in our case it needs the `.leaflet-grab` class), then change the element style.

I think what you want to do now is having another style attribute, something like:
```python
class MapStyle(Style, Widget):
    """ Map Style Widget """
    _model_name = Unicode('LeafletMapStyleModel').tag(sync=True)
    _model_module = Unicode("jupyter-leaflet").tag(sync=True)
    
    default_cursor = Unicode('hand', help='Cursor type').tag(sync=True)
    dragging_cursor = Unicode('move', help='Cursor type').tag(sync=True)
```
And on the JS side something like:
```javascript
LeafletMapStyleModel.styleProperties = {
     default_cursor: {
         selector: '.leaflet-grab',
         attribute: 'cursor',
         default: 'hand'
     },
    dragging_cursor: {
         selector: '.leaflet-dragging',
         attribute: 'cursor',
         default: 'move'
     }
};
```
I guess (hope) it will work

Reply:
but, how will it know that it must be applied when dragging if has no access to `.leaflet-dragging` class?? I'll try though =)

Reply:
I am not sure I understand your question. The `.leaflet-grab` is a CSS class applied by default on the Leaflet map. Now if you inspect the DOM tree with Chrome/Firefox developer tools you see that a `.leaflet-dragging` class is added to the Map element when dragging. So I hope this will work.

Reply:
As per in-person discussion with Sylvain and looking at the code, we think it will not work with the selector `.leaflet-dragging` unfortunately... Because the selector is used once, but it won't know if the `.leaflet-dragging` appears at some point. 

So if you want to implement this logic it must be done in JavaScript. On the `MapView` implementation you can listen for mouse events `mousemove`, and you could change the `style.cursor` value dynamically this way.

Reply:
Ok, I see.. I tried, but didn't work.. as you expected.. do you think that should be changed now? I'll try to do my best, but I cannot guarantee..

Reply:
Other idea: 

On the Python `Map` class you could have two styles:
```python
default_style = InstanceDict(MapStyle).tag(sync=True, **widget_serialization)
dragging_style = InstanceDict(MapStyle, default_value={'cursor': 'move'}).tag(sync=True, **widget_serialization)
``` 

Then on the JavaScript `LeafletMapView` class you could, on mouse events, choose which style is applied.
Something like:
```javascript
leaflet_events: function () {
    var that = this;
    this.obj.on('movestart', function (e) {
        this.model.get('default_style').unstyle();
        this.model.get('dragging_style').style();
    }
    this.obj.on('moveend', function (e) {
        this.model.get('dragging_style').unstyle();
        this.model.get('default_style').style();
    }
    // ...
}
```
I'm using the `style` and `unstyle` method from the style widget here, see https://github.com/jupyter-widgets/ipywidgets/blob/master/packages/base/src/widget_style.ts#L90.

Reply:
It says: `that.model.get(...).unstyle is not a function` :thinking:

Reply:
Oh sorry.. What `that.model.get('default_style')` returns is the `StyleModel`, not a `StyleView`...
Sorry for the time wasted, we are getting there hopefully...

There is a way to get the view from the model. Something like:
```javascript
that.model.get('default_style').views[0].then(function(style_view) {
    style_view.unstyle();
});
```
I start to really feel uncomfortable with those changes though...

Reply:
Oh men, this was a head breaker for me! but I think I got it... 

I forgot about `dragging_style` and only used `style`...
I realized that `view` attribute wasn't an Array but an Object with a unique key...

``` javascript
this.obj.on('moveend', function (e) {
    var style_view = Object.keys(that.model.get('style').views)[0];
    that.model.get('style').views[style_view].then(function(view) {
        view.style();
    });
});
this.obj.on('movestart', function (e) {
    var style_view = Object.keys(that.model.get('style').views)[0];
    that.model.get('style').views[style_view].then(function(view) {
        view.unstyle();
    });
});
```
And this works fine! what do you think??

Reply:
> I forgot about `dragging_style` and only used `style`

I do think we should have a `default_style` and a `dragging_style`. Because here you are using `style` only when dragging, which will not be obvious for other users.

> what do you think??

The fact that we are taking the first style view `Object.keys(that.model.get('style').views)[0]` is a bit weak, as nothing tells us that this first view is the one used by our `Map` (I think if you display the Map multiple times in your Notebook, your solution will work only for the first view of the Map).
Anyway, I'm happy to merge it as it is, and we can fix it later, before making a release.

The way to fix this would be to loop over all the views and call the `(un)style` function only for the view that respects the condition `view.parent === this`.

Reply:
Hi @martinRenou, in my way of changing the code to use a `default_style` I found that when I change the attribute's name from `style` to anything else, it stops working, and inspecting I found that the object is created, but it has no views.. :thinking: maybe is for [this](https://github.com/jupyter-widgets/ipywidgets/blob/cd22809ed525d30ea20f37a65c49de5273de009c/packages/base/src/widget_style.ts#L75) line.. but I am just guessing..
![Pantallazo-2019-04-26 08-08-59](https://user-images.githubusercontent.com/10751505/56804022-b785e080-67fa-11e9-91e5-b200a3c48dee.png)

Reply:
Maybe you forgot to change the name in the deserializer? And in the `LeafletMapModel` property?

Reply:
nope, I changed all

Reply:
Do you mind pushing your changes?

Reply:
Ok, I'll push it with the changes you proposed, with `dragging_style` and `default_style`